pyomo/gdp/plugins/bigm.py

```diff
@@ -2,307 +2,660 @@
+    """
+
+    alias('gdp.bigm', doc=textwrap.fill(textwrap.dedent(__doc__.strip())))
+
+    CONFIG = ConfigBlock("gdp.bigm")
+    CONFIG.declare('bigM', ConfigValue(
+        default=None,
+        domain=_to_dict,
+        description="Big-M value used for constraint relaxation",
+        doc="""
+
+            A user-specified value (or dict) of M values that override
+            M-values found through model Suffixes or that would
+            otherwise be calculated using variable domains."""
+    ))
 
     def __init__(self):
+        """Initialize transformation object."""
         super(BigM_Transformation, self).__init__()
         self.handlers = {
-            Constraint : self._xform_constraint,
-            Var : self._xform_var,
-            Connector : self._xform_skip,
-            Suffix : self._xform_skip,
-            Param : self._xform_skip,
-            }
-
-    def _apply_to(self, instance, **kwds):
-        options = kwds.pop('options', {})
-
-        bigM = options.get('default_bigM', None)
-        bigM = kwds.pop('default_bigM', bigM)
-        if bigM is not None:
-            #
-            # Test for the suffix - this test will (correctly) generate
-            # a warning if the component is already declared, but is a
-            # different ctype (e.g., a constraint or block)
-            #
-            if 'BigM' not in instance.component_map(Suffix):
-                instance.BigM = Suffix(direction=Suffix.LOCAL)
-            #
-            # Note: this will implicitly change the model default BigM
-            # value so that the argument overrides the option, which
-            # overrides any default specified on the model.
-            #
-            instance.BigM[None] = bigM
-
-        targets = kwds.pop('targets', None)
-
-        if kwds:
-            logger.warning("GDP(BigM): unrecognized keyword arguments:\n%s"
-                           % ( '\n'.join(iterkeys(kwds)), ))
+            Constraint:  self._xform_constraint,
+            Var:         False,
+            Connector:   False,
+            Suffix:      False,
+            Param:       False,
+            Set:         False,
+            Disjunction: self._warn_for_active_disjunction,
+            Disjunct:    self._warn_for_active_disjunct,
+            Block:       self._transform_block_on_disjunct,
+        }
+
+    def _get_bigm_suffix_list(self, block):
+        # Note that you can only specify suffixes on BlockData objects or
+        # SimpleBlocks. Though it is possible at this point to stick them
+        # on whatever components you want, we won't pick them up.
+        suffix_list = []
+        while block is not None:
+            bigm = block.component('BigM')
+            if type(bigm) is Suffix:
+                suffix_list.append(bigm)
+            block = block.parent_block()
+        return suffix_list
+
+    def _apply_to(self, instance, targets=None, **kwds):
+        config = self.CONFIG().set_value(kwds.pop('options', {}))
```

Comment:
I'm very confused how this is passing tests, because on my system, `set_value()` has no return value (so it's always `None`), which causes `config = None`, which causes issues when you try to do `config.set_value(kwds)` below.

Reply:
You need the current pyutilib master.

Reply:
Ah, yes. That seems to have done the trick.